### DISCUSSION TITLE
How does one configure the execution backend for maptasks

### DISCUSSION THREAD

**Discussion by palchicz** [flyteorg/flyte#1444] [Q&A] [answered]
Hi Flyte crew, 

I was wondering if I could get some insight into the backend execution of maptasks and how configurable that is. My current understanding is that maptasks use a `container_array` type node as opposed to a `python-task` type node, and that seems to impact how they are run. As far as I can tell, `python-task` nodes look like they are run on a k8 pod whereas I'm not entirely sure how `container_array` nodes get executed (it doesn't seem like a k8 pod though). I also see that one can run maptasks on [aws batch](https://docs.flyte.org/projects/cookbook/en/latest/auto/core/control_flow/map_task.html#sphx-glr-auto-core-control-flow-map-task-py), but it is unclear from the docs how one sets that up. 

So I guess my questions are:
1. What is the default execution backend for maptasks.
2. Can it be changed?
3. What can it be changed to (e.g. k8 or aws batch) and how?

Thanks ahead of time for your response!

**Comment by EngHabu** [comment, marked as answer]
Hi @palchicz,

As you observed, there are two plugins to run maptasks that ship as part of flyteplugins; [k8s array](https://github.com/flyteorg/flyteplugins/blob/master/go/tasks/plugins/array/k8s/executor.go#L174-L182) and [aws array](https://github.com/flyteorg/flyteplugins/blob/master/go/tasks/plugins/array/awsbatch/executor.go#L188-L197).

Which plugin to use is controlled through [this config](https://github.com/flyteorg/flyte/blob/10cee9f139824512b6c5be1667d321bdbc8835fa/charts/flyte/values-sandbox.yaml#L152-L162)

1. By default, as you can see in sandbox, it uses k8s-array plugin. The way this executes array tasks is by launching a pod for every instance in the array. It's a simple and straightforward implementation but has some scale/performance drawbacks that we have some open issues to address ([issue](https://github.com/flyteorg/flyte/issues/1276)) but the big benefit it, it works out of the box.
2. You can implement your own plugin that handles that task type and configure the system to use it instead. 
3. `k8s-array` and `aws_array` are the names of these two plugins. If you change the config, flyte will attempt to use that one to launch array jobs.

The AWS Array Plugin needs some additional configuration to make work which we are in the process of documenting as well. On top of my head, they are:

* You need to configure the AWS Role flytepropeller assumes to be able to `PassRole` all the roles you will use to run jobs in AWS Batch. That requires setting up Trust Relationship
* All workflows/launchplans need to be configured with an `assumable_iam_role` ([example](https://github.com/flyteorg/flytesnacks/blob/28c69268bce3dae574839e7a5923073c18096e35/cookbook/core/sandbox.config#L6)) that flyte will pass to AWS Batch Service to run the jobs as.
* You need to configure AWS Batch Queues and configure FlyteAdmin ([example config](https://github.com/flyteorg/flyteadmin/blob/2d81c1eec24cffb43346b56fc0017fd29db33a38/pkg/runtime/testdata/config.yaml#L1-L20)) to choose which projects/domains/workflows go to which AWS Batch Queus. I strongly recommend, at the very least, separating your development and production domains into separate queues. There is a limitation of the number of queues in AWS Batch so you can't create a queue for every project/domain combination.

cc @migueltol22 as you are going through that process at the moment in case I missed anything?
cc @katrogan, @anandswaminathan if you can correct me or augment my answer if I missed anything?

**Reply by palchicz** [reply]
This is all great (and sorry for the slow response). This should be unblocking for now, but I'll reach out if I come up with any more questions. 

Thanks again!

**Reply by kumare3** [reply]
@palchicz please do and also, join the slack channel where we can help faster

**Reply by EngHabu** [reply]
@palchicz We have now published docs on how to [setup AWS Batch integration](https://docs.flyte.org/en/latest/deployment/plugin_setup/aws/batch.html#deployment-plugin-setup-aws-array). Would love if you give it a shot and let us know...

**Reply by SandraGH5** [reply]
@EngHabu do we need more documentation for this issue?

**Reply by EngHabu** [reply]
This should be all we need.. until customers try it out and give us feedback :-)

**Reply by SandraGH5** [reply]
Sounds good 👍🏻
